server/src/main/java/org/elasticsearch/cluster/metadata/RepositoryMetadata.java

```diff
@@ -37,6 +41,11 @@
      */
     private final long pendingGeneration;
 
+    /**
+     * List of {@link org.elasticsearch.snapshots.SnapshotId} marked as "to delete"
+     */
+    private final List<SnapshotId> snapshotsToDelete;
```

Comment:
I just realised that if we track the pending deletes it here then it'll get lost if someone does a put or delete on this repo, maybe just to adjust its settings or rename it or something. Maybe we should use a separate `Metadata.Custom` instead.

Reply:
I thought I took care of propagating the pending deletes in case of a put on the repository but actually you're right, there's a missing bit I think (see `canUpdateInPlace(newRepositoryMetadata, existing)` in `RepositoriesService`).

Putting the pending deletes in a separate `Custom` also makes sense but we have to decide how to match the repository when it's re-created (by uuid?) and how long should the pending deletes be kept in cluster state. I think it would be easier to keep the pending deletes where it is today and prevent the deletion of a repository if it still have pending deletes (like we prevent repo deletion if there are ongoing operations.

Reply:
++ there's no easy answer. Blocking the removal of the repository would worry me a bit, if we introduced some awful bug that made deletes fail then we could be retrying forever without a way to fix it. I guess we'd also want to prevent marking a repo as readonly for similar reasons. Maybe we'd need a `?force` parameter to the delete repo API as an escape hatch just in case.

Reply:
I refactored things a bit and I added some specific checks for read-only repositories so that we don't trigger snapshot deletions on such repositories. Pending deletes are still added to the repository metadata so that they'll be processed once the repository is writeable again.

Reply:
Also, I did not implement the logic to prevent deleting repository (or making it read-only) when they have pending deletes. I'd like to discuss and address this as a follow up if it is ok for you.